src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -496,11 +526,14 @@ protected HqlTreeNode VisitUnaryExpression(UnaryExpression expression)
 				case ExpressionType.Convert:
 				case ExpressionType.ConvertChecked:
 				case ExpressionType.TypeAs:
-					return IsCastRequired(expression.Operand, expression.Type, out var existType)
+					var notCastable = _notCastableExpressions.TryGetValue(expression, out var castType);
+					castType = castType ?? expression.Type;
```

Comment:
`castType` is null when not `notCastable`. Can this please made explicit?

Reply:
Done.